design/upgrade.md

```diff
@@ -0,0 +1,165 @@
+# **Rook Cluster Upgrades**
+
+## **Overview**
+Over time, new versions with improvements to the Rook software will be released and Rook clusters that have already been deployed should be upgraded to the newly released version.
+Being able to keep the deployed software current is an important part of managing the deployment and ensuring its health.
+In the theme of Rook's orchestration and management capabilities making the life of storage admins easier, this upgrade process should be both automatic and reliable.
+This document will describe a proposed design for the upgrading of Rook software as well as pose questions to the community for feedback so that we can deliver on the goal of automatic and reliable upgrades.
+
+## **Goal**
+In order for software upgrade support in Rook to be considered successful, the goals listed below should be met.
+Note that these goals are for a long term vision and are not all necessarily deliverable within the v0.6 release time frame.
+* **Automatic:** When a new version of Rook is released, a live cluster should be able to update all its components to the new version without user intervention.
+* **No downtime:** During an upgrade window, there should be **zero** downtime of cluster functionality.
+  * The upgrade process should be carried out in a rolling fashion so that not all components are being updated simultaneously.
+The cluster should be maintained in a healthy state the entire time.
+* **Migrations:** Breaking changes, as well as schema and data format changes should be handled through an automated migration processes.
+* **Rollback:** In the event that the upgrade is not successful, the Rook software should be rolled back to the previous version and cluster health should be restored.
+
```

Comment:
Just to be concise with the update scope, i will add a Non-goal section about upgrading ceph.

Also, not sure if we want to talk about it here but we should point out about k8s upgrade and how it affects rook.

Reply:
@kokhang are you saying upgrading ceph is a non-goal? I would disagree. Upgrading ceph and the data path is a very clear goal of the upgrade

Reply:
@travisn sorry.. just to clarify. For this design and feature, we are only focussing on upgrading the Rook controlplane.

Edit: never mind. The rook control plane is also ceph.

Reply:
right, the Rook container packages the ceph daemons/tools as well, so upgrading Rook == upgrading Ceph.